File: src/core/domain/knowledge_learning.py
# =============================================================================
# Domain Layer — Knowledge Learning Engine (FASE 33)
# =============================================================================
# Entidades puras del aprendizaje de conocimiento: runs, etapas, eventos,
# feedback y score de readiness. Sin dependencias externas.
#
# "Learning" NO es fine-tuning de pesos: es descubrimiento de schema, profiling,
# comprensión semántica, validación humana, indexado y evaluación. Los estados
# del pipeline son explícitos y auditables; el LLM nunca es la única fuente de
# verdad (INFERRED != APPROVED).
# =============================================================================
from __future__ import annotations

from dataclasses import dataclass, field
from datetime import datetime, timezone
from enum import StrEnum
from uuid import UUID

# -----------------------------------------------------------------------------
# Estados y etapas
# -----------------------------------------------------------------------------


class LearningRunStatus(StrEnum):
    """Estado del run de aprendizaje (compatible con training_runs)."""

    QUEUED = "queued"
    RUNNING = "running"
    AWAITING_VALIDATION = "awaiting_validation"
    COMPLETED = "completed"
    FAILED = "failed"
    CANCELLED = "cancelled"

    @property
    def is_terminal(self) -> bool:
        return self in (
            LearningRunStatus.COMPLETED,
            LearningRunStatus.FAILED,
            LearningRunStatus.CANCELLED,
        )


class LearningTrigger(StrEnum):
    MANUAL = "manual"
    SCHEDULED = "scheduled"
    ONBOARDING = "onboarding"


class LearningStage(StrEnum):
    """Etapas del pipeline de aprendizaje (explícitas y observables)."""

    CONNECTING = "connecting"
    DISCOVERING_SCHEMA = "discovering_schema"
    PROFILING = "profiling"
    DETECTING_ENTITIES = "detecting_entities"
    ANALYZING_FIELDS = "analyzing_fields"
    DETECTING_RELATIONSHIPS = "detecting_relationships"
    LLM_REASONING = "llm_reasoning"
    GENERATING_QUESTIONS = "generating_questions"
    AWAITING_VALIDATION = "awaiting_validation"
    CHUNKING = "chunking"
    EMBEDDING = "embedding"
    INDEXING = "indexing"
    EVALUATING = "evaluating"
    SCORING = "scoring"
    READY = "ready"


# Etapas que la FASE 1 ejecuta realmente (deterministas). Las fases 2-7
# agregan las restantes a measure que se implementan: nunca se simulan.
PHASE1_ACTIVE_STAGES: tuple[LearningStage, ...] = (
    LearningStage.CONNECTING,
    LearningStage.DISCOVERING_SCHEMA,
    LearningStage.PROFILING,
    LearningStage.DETECTING_ENTITIES,
    LearningStage.ANALYZING_FIELDS,
    LearningStage.DETECTING_RELATIONSHIPS,
    # FASE 33D: preguntas deterministas desde evidencia real (siempre activa).
    LearningStage.GENERATING_QUESTIONS,
    LearningStage.SCORING,
)

# Peso relativo de cada etapa para el progreso global (0-100 real, sin timers).
PHASE1_STAGE_WEIGHTS: dict[LearningStage, int] = {
    LearningStage.CONNECTING: 5,
    LearningStage.DISCOVERING_SCHEMA: 30,
    LearningStage.PROFILING: 15,
    LearningStage.DETECTING_ENTITIES: 15,
    LearningStage.ANALYZING_FIELDS: 20,
    LearningStage.DETECTING_RELATIONSHIPS: 10,
    LearningStage.GENERATING_QUESTIONS: 10,
    LearningStage.EVALUATING: 15,
    LearningStage.SCORING: 5,
    # FASE 33B: solo se incluye cuando el flag LLM está activo.
    LearningStage.LLM_REASONING: 25,
}

# Mapa completo de pesos por etapa (el progreso normaliza sobre las etapas
# realmente incluidas en el run).
STAGE_WEIGHTS: dict[LearningStage, int] = PHASE1_STAGE_WEIGHTS

_STAGE_LABELS: dict[LearningStage, str] = {
    LearningStage.CONNECTING: "Conectando a la fuente",
    LearningStage.DISCOVERING_SCHEMA: "Descubriendo schema",
    LearningStage.PROFILING: "Perfilando columnas",
    LearningStage.DETECTING_ENTITIES: "Detectando entidades",
    LearningStage.ANALYZING_FIELDS: "Analizando campos",
    LearningStage.DETECTING_RELATIONSHIPS: "Entendiendo relaciones",
    LearningStage.LLM_REASONING: "Razonando con el LLM",
    LearningStage.GENERATING_QUESTIONS: "Generando preguntas",
    LearningStage.AWAITING_VALIDATION: "Esperando validación humana",
    LearningStage.CHUNKING: "Fragmentando conocimiento",
    LearningStage.EMBEDDING: "Construyendo embeddings",
    LearningStage.INDEXING: "Indexando",
    LearningStage.EVALUATING: "Evaluando conocimiento",
    LearningStage.SCORING: "Calculando readiness",
    LearningStage.READY: "Conocimiento listo",
}


def stage_label(stage: str | LearningStage) -> str:
    try:
        return _STAGE_LABELS[LearningStage(stage)]
    except (ValueError, KeyError):
        return str(stage)


class StageStatus(StrEnum):
    PENDING = "pending"
    RUNNING = "running"
    COMPLETED = "completed"
    SKIPPED = "skipped"
    FAILED = "failed"


class LLMAnalysisStatus(StrEnum):
    """Estado del análisis LLM por tabla (FASE 33B)."""

    COMPLETED = "completed"
    FAILED = "failed"
    SKIPPED = "skipped"


class QuestionType(StrEnum):
    """Origen/tipo de una pregunta de negocio (FASE 33D)."""

    ENUM_MEANING = "enum_meaning"
    FIELD_AMBIGUITY = "field_ambiguity"
    TABLE_VARIANT = "table_variant"
    TAX_INCLUSION = "tax_inclusion"
    RELATIONSHIP_MEANING = "relationship_meaning"
    BUSINESS_RULE = "business_rule"
    METRIC_DEFINITION = "metric_definition"
    LLM_QUESTION = "llm_question"
    OTHER = "other"


class QuestionPriority(StrEnum):
    CRITICAL = "critical"
    HIGH = "high"
    MEDIUM = "medium"
    LOW = "low"


class QuestionStatus(StrEnum):
    PENDING = "pending"
    ANSWERED = "answered"
    SKIPPED = "skipped"
    DEFERRED = "deferred"
    EXPIRED = "expired"


class FeedbackSource(StrEnum):
    QUESTION = "question"
    REVIEW_QUEUE = "review_queue"
    STUDIO = "studio"
    IMPORT = "import"


QUESTION_PRIORITY_WEIGHTS: dict[str, float] = {
    "ambiguity": 0.30,
    "business_impact": 0.25,
    "retrieval_impact": 0.15,
    "frequency": 0.10,
    "confidence_gap": 0.10,
    "dependents": 0.10,
}


def _clamp01(value: float) -> float:
    return max(0.0, min(1.0, float(value)))


def compute_question_priority(
    *,
    ambiguity: float,
    business_impact: float,
    retrieval_impact: float,
    frequency: float,
    confidence: float,
    dependents: float,
) -> tuple[QuestionPriority, float]:
    """Score de prioridad 0..1 explicable (sección 8) y etiqueta.

    No bombardea al cliente: solo CRITICAL/HIGH deberían bloquear el gate.
    """
    signals = {
        "ambiguity": _clamp01(ambiguity),
        "business_impact": _clamp01(business_impact),
        "retrieval_impact": _clamp01(retrieval_impact),
        "frequency": _clamp01(frequency),
        "confidence_gap": 1.0 - _clamp01(confidence),
        "dependents": _clamp01(dependents),
    }
    total_weight = sum(QUESTION_PRIORITY_WEIGHTS.values())
    score = round(
        sum(
            signals[key] * weight
            for key, weight in QUESTION_PRIORITY_WEIGHTS.items()
        )
        / total_weight,
        4,
    )
    if score >= 0.78:
        return QuestionPriority.CRITICAL, score
    if score >= 0.58:
        return QuestionPriority.HIGH, score
    if score >= 0.38:
        return QuestionPriority.MEDIUM, score
    return QuestionPriority.LOW, score


class KnowledgeGate(StrEnum):
    """Knowledge Readiness Gate (sección 31)."""

    NOT_READY = "NOT_READY"
    LEARNING = "LEARNING"
    NEEDS_INPUT = "NEEDS_INPUT"
    READY = "READY"
    DEGRADED = "DEGRADED"


class KnowledgeEventType(StrEnum):
    STAGE_STARTED = "stage.started"
    STAGE_COMPLETED = "stage.completed"
    STAGE_FAILED = "stage.failed"
    LEARNING_STARTED = "learning.started"
    SCHEMA_DISCOVERED = "schema.discovered"
    TABLE_ANALYZED = "table.analyzed"
    ENTITY_DETECTED = "entity.detected"
    FIELD_DETECTED = "field.detected"
    RELATIONSHIP_DETECTED = "relationship.detected"
    LLM_ANALYSIS_STARTED = "llm.analysis.started"
    LLM_ANALYSIS_COMPLETED = "llm.analysis.completed"
    QUESTION_GENERATED = "question.generated"
    KNOWLEDGE_CONFIRMED = "knowledge.confirmed"
    EMBEDDING_STARTED = "embedding.started"
    EMBEDDING_PROGRESS = "embedding.progress"
    EVALUATION_STARTED = "evaluation.started"
    EVALUATION_COMPLETED = "evaluation.completed"
    SCORE_COMPUTED = "knowledge.score.computed"
    LEARNING_COMPLETED = "learning.completed"
    LEARNING_FAILED = "learning.failed"
    LEARNING_CANCELLED = "learning.cancelled"


class EventCategory(StrEnum):
    """Filtros del Activity Feed (All / Discovery / AI / Validation / Indexing)."""

    DISCOVERY = "discovery"
    AI = "ai"
    VALIDATION = "validation"
    INDEXING = "indexing"
    SYSTEM = "system"


_CATEGORY_BY_EVENT: dict[str, EventCategory] = {
    KnowledgeEventType.LEARNING_STARTED.value: EventCategory.SYSTEM,
    KnowledgeEventType.LEARNING_COMPLETED.value: EventCategory.SYSTEM,
    KnowledgeEventType.LEARNING_FAILED.value: EventCategory.SYSTEM,
    KnowledgeEventType.LEARNING_CANCELLED.value: EventCategory.SYSTEM,
    KnowledgeEventType.STAGE_STARTED.value: EventCategory.SYSTEM,
    KnowledgeEventType.STAGE_COMPLETED.value: EventCategory.SYSTEM,
    KnowledgeEventType.STAGE_FAILED.value: EventCategory.SYSTEM,
    KnowledgeEventType.SCHEMA_DISCOVERED.value: EventCategory.DISCOVERY,
    KnowledgeEventType.TABLE_ANALYZED.value: EventCategory.DISCOVERY,
    KnowledgeEventType.ENTITY_DETECTED.value: EventCategory.AI,
    KnowledgeEventType.FIELD_DETECTED.value: EventCategory.AI,
    KnowledgeEventType.RELATIONSHIP_DETECTED.value: EventCategory.AI,
    KnowledgeEventType.LLM_ANALYSIS_STARTED.value: EventCategory.AI,
    KnowledgeEventType.LLM_ANALYSIS_COMPLETED.value: EventCategory.AI,
    KnowledgeEventType.QUESTION_GENERATED.value: EventCategory.VALIDATION,
    KnowledgeEventType.KNOWLEDGE_CONFIRMED.value: EventCategory.VALIDATION,
    KnowledgeEventType.EMBEDDING_STARTED.value: EventCategory.INDEXING,
    KnowledgeEventType.EMBEDDING_PROGRESS.value: EventCategory.INDEXING,
    KnowledgeEventType.EVALUATION_STARTED.value: EventCategory.INDEXING,
    KnowledgeEventType.EVALUATION_COMPLETED.value: EventCategory.INDEXING,
    KnowledgeEventType.SCORE_COMPUTED.value: EventCategory.SYSTEM,
}


def event_category(event_type: str) -> EventCategory:
    return _CATEGORY_BY_EVENT.get(event_type, EventCategory.SYSTEM)


# -----------------------------------------------------------------------------
# Score (Knowledge Readiness, sección 11)
# -----------------------------------------------------------------------------

DEFAULT_SCORE_WEIGHTS: dict[str, float] = {
    "schema_discovery": 0.15,
    "semantic_understanding": 0.20,
    "relationships": 0.15,
    "human_validation": 0.15,
    "knowledge_coverage": 0.15,
    "rag_evaluation": 0.15,
    "data_freshness": 0.05,
}

DEFAULT_GATE_THRESHOLDS: dict[str, float] = {
    # Score global mínimo para READY.
    "ready_overall": 80.0,
    # Umbrales por dimensión exigidos por el gate READY (sección 31).
    "schema_coverage": 90.0,
    "semantic_coverage": 70.0,
    "evaluation_score": 75.0,
    # Preguntas críticas pendientes máximas para READY.
    "critical_questions_max": 0,
    # Debajo de esto el conocimiento está "learning"/no listo.
    "min_overall": 40.0,
    # Días sin escanear para considerar la fuente stale (DEGRADED).
    "stale_source_days": 14.0,
}


@dataclass(kw_only=True)
class ScoreDimension:
    """Una dimensión del score con composición explicable."""

    key: str
    label: str
    score: float  # 0-100
    weight: float  # 0-1
    detail: str = ""
    measured: bool = True

    def to_dict(self) -> dict:
        return {
            "key": self.key,
            "label": self.label,
            "score": round(self.score, 2),
            "weight": round(self.weight, 4),
            "detail": self.detail,
            "measured": self.measured,
        }


@dataclass(kw_only=True)
class KnowledgeScoreResult:
    """Resultado de Knowledge Readiness con razones accionables."""

    organization_id: UUID
    source_id: UUID | None = None
    run_id: UUID | None = None
    overall: float = 0.0
    gate: KnowledgeGate = KnowledgeGate.NOT_READY
    dimensions: list[ScoreDimension] = field(default_factory=list)
    reasons: list[str] = field(default_factory=list)
    weights: dict[str, float] = field(default_factory=dict)
    computed_at: datetime = field(default_factory=lambda: datetime.now(timezone.utc))

    def dimension(self, key: str) -> ScoreDimension | None:
        return next((d for d in self.dimensions if d.key == key), None)

    def to_dict(self) -> dict:
        return {
            "organization_id": str(self.organization_id),
            "source_id": str(self.source_id) if self.source_id else None,
            "run_id": str(self.run_id) if self.run_id else None,
            "overall": round(self.overall, 2),
            "gate": self.gate.value,
            "dimensions": [d.to_dict() for d in self.dimensions],
            "reasons": list(self.reasons),
            "weights": {k: round(v, 4) for k, v in self.weights.items()},
            "computed_at": self.computed_at.isoformat(),
        }


def compute_overall(
    scores: dict[str, float], weights: dict[str, float]
) -> float:
    """Promedio ponderado normalizado (dimensiones ausentes = 0)."""
    total_weight = sum(max(w, 0.0) for w in weights.values())
    if total_weight <= 0:
        return 0.0
    acc = sum(
        max(0.0, min(100.0, scores.get(key, 0.0))) * max(weight, 0.0)
        for key, weight in weights.items()
    )
    return round(acc / total_weight, 2)


# -----------------------------------------------------------------------------
# Runs / steps / events / feedback
# -----------------------------------------------------------------------------


@dataclass(kw_only=True)
class KnowledgeLearningRun:
    id: UUID
    organization_id: UUID
    catalog_source_id: UUID | None = None
    workspace_id: UUID | None = None
    kb_source_id: UUID | None = None
    training_run_id: UUID | None = None
    trigger: LearningTrigger = LearningTrigger.MANUAL
    status: LearningRunStatus = LearningRunStatus.QUEUED
    current_stage: str = LearningStage.CONNECTING.value
    overall_progress: int = 0
    stage_progress: int = 0
    gate: KnowledgeGate | None = None
    tables_analyzed: int = 0
    entities_detected: int = 0
    fields_detected: int = 0
    relationships_detected: int = 0
    metrics: dict = field(default_factory=dict)
    error_summary: dict = field(default_factory=dict)
    created_by: UUID | None = None
    created_at: datetime | None = None
    started_at: datetime | None = None
    finished_at: datetime | None = None


@dataclass(kw_only=True)
class KnowledgeLearningStep:
    id: UUID
    organization_id: UUID
    run_id: UUID
    stage: str
    sequence: int = 0
    status: StageStatus = StageStatus.PENDING
    progress: int = 0
    metrics: dict = field(default_factory=dict)
    error: str | None = None
    started_at: datetime | None = None
    finished_at: datetime | None = None
    duration_ms: float = 0.0


@dataclass(kw_only=True)
class KnowledgeEvent:
    id: UUID
    organization_id: UUID
    event_type: str
    run_id: UUID | None = None
    source_id: UUID | None = None
    stage: str | None = None
    category: str = EventCategory.SYSTEM.value
    severity: str = "info"  # info | success | warning | error
    message: str = ""
    payload: dict = field(default_factory=dict)
    seq: int = 0
    created_at: datetime | None = None


@dataclass(kw_only=True)
class KnowledgeLearningSettings:
    """Configuración por tenant: pesos del score y umbrales del gate."""

    organization_id: UUID
    weights: dict[str, float] = field(
        default_factory=lambda: dict(DEFAULT_SCORE_WEIGHTS)
    )
    thresholds: dict[str, float] = field(
        default_factory=lambda: dict(DEFAULT_GATE_THRESHOLDS)
    )
    updated_by: UUID | None = None
    updated_at: datetime | None = None

    def effective_weights(self) -> dict[str, float]:
        merged = dict(DEFAULT_SCORE_WEIGHTS)
        for key, value in (self.weights or {}).items():
            if key in merged and isinstance(value, (int, float)) and value >= 0:
                merged[key] = float(value)
        return merged

    def effective_thresholds(self) -> dict[str, float]:
        merged = dict(DEFAULT_GATE_THRESHOLDS)
        for key, value in (self.thresholds or {}).items():
            if key in merged and isinstance(value, (int, float)):
                merged[key] = float(value)
        return merged

File: portal/src/lib/knowledgeLearning.ts
// =============================================================================
// Knowledge Learning API — tipos reales y streaming SSE (FASE 33E)
// =============================================================================
// Todo lo que muestra la UI proviene de estados reales del backend:
// /status, /sources, /runs/{id}, /runs/{id}/events (SSE), /questions,
// /entities y /score. Nunca se inventa progreso ni confianza.
// =============================================================================
import { api, loadSession } from "../api";
import { emitAuthExpired } from "./errors";

function withSession<T>(
  path: string,
  options: RequestInit = {}
): Promise<T> {
  const session = loadSession();
  return api<T>(path, {
    ...options,
    token: session?.token,
    organizationId: session?.organizationId,
  });
}

export type KnowledgeGate =
  | "NOT_READY"
  | "LEARNING"
  | "NEEDS_INPUT"
  | "READY"
  | "DEGRADED";

export type ScoreDimension = {
  key: string;
  label: string;
  score: number;
  weight: number;
  detail: string;
  measured: boolean;
};

export type KnowledgeScore = {
  overall: number;
  gate: KnowledgeGate;
  dimensions: ScoreDimension[];
  reasons: string[];
  weights: Record<string, number>;
  computed_at: string | null;
  source_id?: string | null;
};

export type LearningStatusCounts = {
  sources_connected: number;
  tables_total: number;
  columns_total: number;
  entities_total: number;
  entities_understood: number;
  fields_total: number;
  relationships_total: number;
  relationships_confirmed: number;
  pending_questions: number;
  verified_knowledge: number;
  open_gaps: number;
};

export type LearningStatus = {
  enabled: boolean;
  headline: string;
  counts: LearningStatusCounts;
  readiness: { overall: number | null; gate: KnowledgeGate; computed_at: string | null };
  active_runs: LearningRun[];
  sources: SourceLearning[];
};

export type SourceLearning = {
  source_id: string;
  connector_id: string;
  engine: string | null;
  connectivity: "healthy" | "degraded" | "unknown";
  phase: string | null;
  last_learning_at: string | null;
  scan_error: string | null;
  knowledge: { overall: number | null; gate: KnowledgeGate };
  tables: { analyzed: number; total: number };
  fields: { understood: number; total: number; documented: number };
  relationships: { discovered: number; confirmed: number };
  pending_questions: number;
  active_run: {
    id: string;
    status: string;
    current_stage: string;
    overall_progress: number;
  } | null;
};

export type LearningRun = {
  id: string;
  organization_id: string;
  catalog_source_id: string | null;
  status: "queued" | "running" | "awaiting_validation" | "completed" | "failed" | "cancelled";
  current_stage: string;
  overall_progress: number;
  stage_progress: number;
  gate: KnowledgeGate | null;
  tables_analyzed: number;
  entities_detected: number;
  fields_detected: number;
  relationships_detected: number;
  metrics: Record<string, unknown>;
  error_summary: Record<string, unknown>;
  created_at: string | null;
  started_at: string | null;
  finished_at: string | null;
  steps?: LearningStep[];
};

export type LearningStep = {
  id: string;
  stage: string;
  sequence: number;
  status: "pending" | "running" | "completed" | "skipped" | "failed";
  progress: number;
  metrics: Record<string, unknown>;
  error: string | null;
  started_at: string | null;
  finished_at: string | null;
  duration_ms: number;
};

export type LearningEvent = {
  id: string;
  seq: number;
  run_id: string | null;
  source_id: string | null;
  event_type: string;
  stage: string | null;
  category: "discovery" | "ai" | "validation" | "indexing" | "system";
  severity: "info" | "success" | "warning" | "error";
  message: string;
  payload: Record<string, unknown>;
  created_at: string | null;
};

export type KnowledgeQuestionOption = {
  value: string;
  label: string;
  occurrence_count?: number;
};

export type KnowledgeQuestion = {
  id: string;
  source_id: string | null;
  run_id: string | null;
  entity_id: string | null;
  field_id: string | null;
  column_id: string | null;
  question_type: string;
  title: string;
  body: string;
  evidence: string[];
  options: KnowledgeQuestionOption[];
  answer_schema: Record<string, unknown>;
  priority: "critical" | "high" | "medium" | "low";
  priority_score: number;
  impact: Record<string, unknown>;
  status: "pending" | "answered" | "skipped" | "deferred" | "expired";
  answer: Record<string, unknown>;
  structured_answer: Record<string, unknown>;
  confidence_before: number | null;
  confidence_after: number | null;
  created_at: string | null;
  answered_at: string | null;
};

export type LearnedEntity = {
  entity_id: string;
  name: string;
  display_name: string;
  description: string | null;
  confidence: number;
  confidence_label: string;
  provenance: string;
  status: string;
  table: string | null;
  source_id: string | null;
  fields_total: number;
  fields_understood: number;
  columns_total: number;
  coverage_pct: number | null;
  relationships_total: number;
  relationships_confirmed: number;
  business_rules_total: number;
  business_rules_approved: number;
  open_questions: number;
  last_learned_at: string | null;
};

export const GATE_LABELS: Record<KnowledgeGate, string> = {
  NOT_READY: "Aún no listo",
  LEARNING: "Aprendiendo",
  NEEDS_INPUT: "Necesita tu ayuda",
  READY: "Listo",
  DEGRADED: "Degradado",
};

export const STAGE_LABELS: Record<string, string> = {
  connecting: "Conectando a la fuente",
  discovering_schema: "Descubriendo schema",
  profiling: "Perfilando columnas",
  detecting_entities: "Detectando entidades",
  analyzing_fields: "Analizando campos",
  detecting_relationships: "Entendiendo relaciones",
  llm_reasoning: "Razonando con el LLM",
  generating_questions: "Generando preguntas",
  awaiting_validation: "Esperando tu validación",
  chunking: "Fragmentando conocimiento",
  embedding: "Construyendo embeddings",
  indexing: "Indexando",
  evaluating: "Evaluando conocimiento",
  scoring: "Calculando readiness",
  ready: "Conocimiento listo",
};

export const CATEGORY_LABELS: Record<string, string> = {
  discovery: "Descubrimiento",
  ai: "IA",
  validation: "Validación",
  indexing: "Indexado",
  system: "Sistema",
};

export function stageLabel(stage: string | null | undefined): string {
  if (!stage) return "";
  return STAGE_LABELS[stage] || stage;
}

export function priorityTone(priority: string): string {
  if (priority === "critical") return "badge-danger";
  if (priority === "high") return "badge-pending";
  if (priority === "medium") return "badge-muted";
  return "badge-muted";
}

export function gateTone(gate: KnowledgeGate | null | undefined): string {
  if (!gate) return "badge-muted";
  if (gate === "READY") return "badge-ok";
  if (gate === "NEEDS_INPUT") return "badge-pending";
  if (gate === "DEGRADED") return "badge-danger";
  if (gate === "LEARNING") return "badge-pending";
  return "badge-muted";
}

// ---------------------------------------------------------------------------
// Fetchers
// ---------------------------------------------------------------------------

export function fetchLearningStatus(): Promise<LearningStatus> {
  return withSession<LearningStatus>("/api/v1/knowledge/learning/status");
}

export function fetchLearningSources(): Promise<SourceLearning[]> {
  return withSession<SourceLearning[]>("/api/v1/knowledge/learning/sources");
}

export function fetchLearningRun(runId: string): Promise<LearningRun> {
  return withSession<LearningRun>(`/api/v1/knowledge/learning/runs/${runId}`);
}

export function fetchRunEvents(
  runId: string,
  sinceSeq = 0,
  limit = 500
): Promise<{ events: LearningEvent[]; count: number }> {
  return withSession<{ events: LearningEvent[]; count: number }>(
    `/api/v1/knowledge/learning/runs/${runId}/events?since_seq=${sinceSeq}&limit=${limit}`
  );
}

export function fetchEvents(
  params: { category?: string; source_id?: string; limit?: number } = {}
): Promise<{ events: LearningEvent[]; count: number }> {
  const query = new URLSearchParams();
  if (params.category) query.set("category", params.category);
  if (params.source_id) query.set("source_id", params.source_id);
  query.set("limit", String(params.limit ?? 80));
  return withSession<{ events: LearningEvent[]; count: number }>(
    `/api/v1/knowledge/learning/events?${query.toString()}`
  );
}

export function fetchQuestions(
  params: { source_id?: string; status?: string; entity_id?: string; limit?: number } = {}
): Promise<{
  questions: KnowledgeQuestion[];
  count: number;
  pending: number;
  blocking: number;
}> {
  const query = new URLSearchParams();
  if (params.source_id) query.set("source_id", params.source_id);
  if (params.entity_id) query.set("entity_id", params.entity_id);
  query.set("status", params.status ?? "pending");
  query.set("limit", String(params.limit ?? 100));
  return withSession(`/api/v1/knowledge/learning/questions?${query.toString()}`);
}

export function fetchLearnedEntities(
  sourceId?: string,
  limit = 24
): Promise<{ entities: LearnedEntity[]; count: number }> {
  const query = new URLSearchParams();
  if (sourceId) query.set("source_id", sourceId);
  query.set("limit", String(limit));
  return withSession<{ entities: LearnedEntity[]; count: number }>(
    `/api/v1/knowledge/learning/entities?${query.toString()}`
  );
}

export function fetchKnowledgeScore(sourceId?: string): Promise<KnowledgeScore> {
  const query = sourceId ? `?source_id=${sourceId}` : "";
  return withSession<KnowledgeScore>(`/api/v1/knowledge/learning/score${query}`);
}

export function startLearning(catalogSourceId: string): Promise<{
  run: LearningRun;
  job_id: string;
}> {
  return withSession("/api/v1/knowledge/learning/start", {
    method: "POST",
    body: JSON.stringify({ catalog_source_id: catalogSourceId }),
  });
}

export function cancelLearning(runId: string): Promise<{ cancelled: string }> {
  return withSession(`/api/v1/knowledge/learning/runs/${runId}/cancel`, {
    method: "POST",
  });
}

export function answerQuestion(
  questionId: string,
  payload: { answer?: string; structured_answer?: Record<string, unknown>; choice?: string }
): Promise<{ question: KnowledgeQuestion; applied_to: Array<Record<string, unknown>> }> {
  return withSession(`/api/v1/knowledge/learning/questions/${questionId}/answer`, {
    method: "POST",
    body: JSON.stringify(payload),
  });
}

export function skipQuestion(
  questionId: string,
  reason?: string
): Promise<{ question: KnowledgeQuestion }> {
  return withSession(`/api/v1/knowledge/learning/questions/${questionId}/skip`, {
    method: "POST",
    body: JSON.stringify({ reason: reason ?? null }),
  });
}

export function deferQuestion(questionId: string): Promise<{ question: KnowledgeQuestion }> {
  return withSession(`/api/v1/knowledge/learning/questions/${questionId}/defer`, {
    method: "POST",
    body: JSON.stringify({}),
  });
}

// ---------------------------------------------------------------------------
// SSE: fetch reader (EventSource no soporta headers Authorization)
// ---------------------------------------------------------------------------

export type StreamHandle = { close: () => void };

export function streamRunEvents(options: {
  runId: string;
  sinceSeq?: number;
  onEvent: (event: LearningEvent) => void;
  onOpen?: () => void;
  onError?: (message: string) => void;
}): StreamHandle {
  const controller = new AbortController();
  const session = loadSession();
  const token = session?.token;
  const organizationId = session?.organizationId;

  (async () => {
    try {
      const headers: Record<string, string> = { Accept: "text/event-stream" };
      if (token) headers.Authorization = `Bearer ${token}`;
      if (organizationId) headers["X-Organization-Id"] = organizationId;
      const response = await fetch(
        `/api/v1/knowledge/learning/runs/${options.runId}/stream?since_seq=${
          options.sinceSeq ?? 0
        }`,
        { headers, signal: controller.signal, credentials: "same-origin" }
      );
      if (response.status === 401) {
        emitAuthExpired("tenant");
        options.onError?.("stream_http_401");
        return;
      }
      if (!response.ok || !response.body) {
        options.onError?.(`stream_http_${response.status}`);
        return;
      }
      options.onOpen?.();
      const reader = response.body.getReader();
      const decoder = new TextDecoder();
      let buffer = "";
      for (;;) {
        const { done, value } = await reader.read();
        if (done) break;
        buffer += decoder.decode(value, { stream: true });
        const frames = buffer.split("\n\n");
        buffer = frames.pop() ?? "";
        for (const frame of frames) {
          const eventLine = frame
            .split("\n")
            .find((line) => line.startsWith("event: "));
          const dataLine = frame.split("\n").find((line) => line.startsWith("data: "));
          if (!dataLine || !eventLine) continue;
          const eventType = eventLine.slice(7).trim();
          if (eventType === "heartbeat") continue;
          try {
            const payload = JSON.parse(dataLine.slice(6)) as LearningEvent;
            options.onEvent({ ...payload, event_type: payload.event_type || eventType });
          } catch {
            // frame incompleto: ignorar
          }
        }
      }
    } catch (error) {
      if (!controller.signal.aborted) {
        options.onError?.(error instanceof Error ? error.message : "stream_error");
      }
    }
  })();

  return { close: () => controller.abort() };
}
